In [1]:
from tensorflow.keras.datasets import imdb
(train_data, train_labels), (test_data, test_labels) = imdb.load_data(num_words=10000)

17464789/17464789 [==============================] - 2s 0us/step


In [2]:
import numpy as np

def vectorize_sequences(sequences, dimension=10000):
  results = np.zeros((len(sequences), dimension))
  for i, sequence in enumerate(sequences):
    for j in sequence:
      results[i, j] = 1.
  return results

x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

In [3]:
y_train = np.asarray(train_labels).astype("float32")
y_test = np.asarray(test_labels).astype("float32")

In [4]:
x_val = x_train[:10000]
partial_x_train = x_train[10000:]
y_val = y_train[:10000]
partial_y_train = y_train[10000:]

In [5]:
from tensorflow import keras
from tensorflow.keras import layers

model = keras.Sequential([
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])

model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=4,
                    batch_size=512,
                    validation_data=(x_val, y_val))

Epoch 1/4
30/30 [==============================] - 4s 69ms/step - loss: 0.5800 - accuracy: 0.7379 - val_loss: 0.4311 - val_accuracy: 0.8629
Epoch 2/4
30/30 [==============================] - 1s 24ms/step - loss: 0.3511 - accuracy: 0.8840 - val_loss: 0.3192 - val_accuracy: 0.8838
Epoch 3/4
30/30 [==============================] - 1s 25ms/step - loss: 0.2527 - accuracy: 0.9167 - val_loss: 0.2875 - val_accuracy: 0.8850
Epoch 4/4
30/30 [==============================] - 1s 22ms/step - loss: 0.1962 - accuracy: 0.9339 - val_loss: 0.3223 - val_accuracy: 0.8720


In [6]:
model.evaluate(x_test, y_test)

782/782 [==============================] - 2s 3ms/step - loss: 0.3440 - accuracy: 0.8595


[0.34403252601623535, 0.8594800233840942]

Overall, we would choose a model with one intermediate layer as overall losses are lower and accuracies are higher.

## **Changing the number of units**

We now try different numbers of units in our intermediate layer. We start with 32 and then 64.

In [7]:
model = keras.Sequential([
    layers.Dense(32, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])

model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

Epoch 1/20
30/30 [==============================] - 4s 102ms/step - loss: 0.4890 - accuracy: 0.7931 - val_loss: 0.3675 - val_accuracy: 0.8746
Epoch 2/20
30/30 [==============================] - 1s 23ms/step - loss: 0.3076 - accuracy: 0.8973 - val_loss: 0.3071 - val_accuracy: 0.8860
Epoch 3/20
30/30 [==============================] - 1s 22ms/step - loss: 0.2429 - accuracy: 0.9177 - val_loss: 0.2848 - val_accuracy: 0.8886
Epoch 4/20
30/30 [==============================] - 1s 22ms/step - loss: 0.2050 - accuracy: 0.9309 - val_loss: 0.2786 - val_accuracy: 0.8882
Epoch 5/20
30/30 [==============================] - 1s 23ms/step - loss: 0.1814 - accuracy: 0.9395 - val_loss: 0.2748 - val_accuracy: 0.8866
Epoch 6/20
30/30 [==============================] - 1s 24ms/step - loss: 0.1587 - accuracy: 0.9463 - val_loss: 0.2823 - val_accuracy: 0.8869
Epoch 7/20
30/30 [==============================] - 1s 22ms/step - loss: 0.1436 - accuracy: 0.9549 - val_loss: 0.3028 - val_accuracy: 0.8781
Epoch 8/20
3

In [8]:
model = keras.Sequential([
    layers.Dense(32, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])

model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=5,
                    batch_size=512,
                    validation_data=(x_val, y_val))

Epoch 1/5
30/30 [==============================] - 3s 67ms/step - loss: 0.5039 - accuracy: 0.7787 - val_loss: 0.3804 - val_accuracy: 0.8683
Epoch 2/5
30/30 [==============================] - 1s 38ms/step - loss: 0.3180 - accuracy: 0.8954 - val_loss: 0.3133 - val_accuracy: 0.8830
Epoch 3/5
30/30 [==============================] - 1s 32ms/step - loss: 0.2519 - accuracy: 0.9164 - val_loss: 0.2871 - val_accuracy: 0.8910
Epoch 4/5
30/30 [==============================] - 1s 27ms/step - loss: 0.2119 - accuracy: 0.9295 - val_loss: 0.3008 - val_accuracy: 0.8777
Epoch 5/5
30/30 [==============================] - 1s 21ms/step - loss: 0.1874 - accuracy: 0.9372 - val_loss: 0.2749 - val_accuracy: 0.8867


In [9]:
 model.evaluate(x_test, y_test)

782/782 [==============================] - 3s 3ms/step - loss: 0.2872 - accuracy: 0.8833


[0.28721413016319275, 0.8832799792289734]

In [10]:
model = keras.Sequential([
    layers.Dense(64, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])

model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

Epoch 1/20
30/30 [==============================] - 3s 64ms/step - loss: 0.4952 - accuracy: 0.7689 - val_loss: 0.3642 - val_accuracy: 0.8652
Epoch 2/20
30/30 [==============================] - 1s 21ms/step - loss: 0.3015 - accuracy: 0.8937 - val_loss: 0.3090 - val_accuracy: 0.8795
Epoch 3/20
30/30 [==============================] - 1s 22ms/step - loss: 0.2405 - accuracy: 0.9158 - val_loss: 0.3013 - val_accuracy: 0.8761
Epoch 4/20
30/30 [==============================] - 1s 22ms/step - loss: 0.2125 - accuracy: 0.9248 - val_loss: 0.3055 - val_accuracy: 0.8751
Epoch 5/20
30/30 [==============================] - 1s 22ms/step - loss: 0.1819 - accuracy: 0.9385 - val_loss: 0.2804 - val_accuracy: 0.8877
Epoch 6/20
30/30 [==============================] - 1s 23ms/step - loss: 0.1648 - accuracy: 0.9438 - val_loss: 0.2784 - val_accuracy: 0.8858
Epoch 7/20
30/30 [==============================] - 1s 23ms/step - loss: 0.1429 - accuracy: 0.9541 - val_loss: 0.3015 - val_accuracy: 0.8827
Epoch 8/20
30

In [11]:
model = keras.Sequential([
    layers.Dense(64, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])

model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])

history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=5,
                    batch_size=512,
                    validation_data=(x_val, y_val))

Epoch 1/5
30/30 [==============================] - 2s 64ms/step - loss: 0.4935 - accuracy: 0.7770 - val_loss: 0.3637 - val_accuracy: 0.8701
Epoch 2/5
30/30 [==============================] - 1s 25ms/step - loss: 0.3034 - accuracy: 0.8910 - val_loss: 0.3007 - val_accuracy: 0.8851
Epoch 3/5
30/30 [==============================] - 1s 40ms/step - loss: 0.2424 - accuracy: 0.9149 - val_loss: 0.2909 - val_accuracy: 0.8838
Epoch 4/5
30/30 [==============================] - 1s 29ms/step - loss: 0.2034 - accuracy: 0.9305 - val_loss: 0.2886 - val_accuracy: 0.8834
Epoch 5/5
30/30 [==============================] - 1s 21ms/step - loss: 0.1794 - accuracy: 0.9387 - val_loss: 0.2898 - val_accuracy: 0.8826


A model with one intermediate layer and 32 units yields better results with a test loss of 0.287 and an accuracy of 0.88.